packages/ERTP/src/paymentLedger.js

```diff
@@ -291,7 +291,22 @@ export const makePaymentLedger = (
    */
   const withdraw = (currentBalance, updatePurseBalance, amount) => {
     amount = coerce(amount);
-    const newPurseBalance = subtract(currentBalance, amount);
+    let newPurseBalance;
+    try {
+      newPurseBalance = subtract(currentBalance, amount);
+    } catch (err) {
+      if (err instanceof RangeError) {
+        const withdrawalError = Error(
+          // @ts-ignore `X` (aka `details`) lazily constructs a
+          // string, but the types do not reflect that
+          X`Withdrawal of ${amount} failed because the purse only contained ${currentBalance}`,
```

Comment:
```suggestion
          X`Withdrawal of ${q(amount)} failed because the purse only contained ${q(currentBalance)}`,
```

I think this captures your intent. Will need to grab `const { quote: q, details: X } = assert`.

Reply:
Hmm, I don't think we want to `quote` the amounts. On one hand, I'm very skeptical of the business value of hiding the any data from the console and selectively revealing it at this point, but on the other hand, if that's what we've decided to do as a company, I think the `amount` to be withdrawn is more sensitive than pretty much anything else we have in vats at this point.

Reply:
The test case below suggests that the amounts are getting quoted anyway. I reread `ses-ava` to see whether we made `t.throws` test against what the `message` would have been if every detail were quoted.

Reply:
Well, rereading `ses-ava` was insightful. It does nothing special to `t.throws` so the message you’re testing below is as explicit as if `q()` were used here. Please disregard my suggestion since you don’t want the quoted amount to appear in the message, but we need something else here for the desired effect and I don’t know what it is.

Reply:
Ah, we got a tip from @erights. Changing `new Error` to `assert.error` will hide the amount from intermediate stack frames. It will also hide the amount from `t.throws`, so the message in the assertion below would need to redact the amount to pass when this works.

Reply:
> ... I'm very skeptical of the business value of hiding any data from the console and selectively revealing it at this point ...

We don't (aim to) hide data from the console; we aim to hide it from parties between the thrower and the console; that is: we aim to keep callers from using `catch` to get unauthorized access to capabilities thrown in exceptions.

Section 4.3 **Exceptions and Errors** of [the 2010 Joe-E paper](https://people.eecs.berkeley.edu/~daw/papers/joe-e-ndss10.pdf) is a good write-up, including this example:

> To see how this can cause unpleasant surprises, suppose Alice calls Bob. Bob has some special capability that she
lacks, and Bob wants to avoid leaking this to her. At some point, Bob might need to invoke Chuck to perform some operation, passing this capability to Chuck. If (unbeknownst to Bob) Chuck can throw an exception that Bob doesn’t catch, this exception might propagate to Alice. If this exception contains Bob’s precious capability, this might cause the capability to leak to Alice, against Bob’s wishes and despite Chuck’s good intentions. 


See also [exception unsealer to go with SES console · Issue \#731 · endojs/endo](https://github.com/endojs/endo/issues/731)

Reply:
> Hmm, I don't think we want to quote the amounts. On one hand, I'm very skeptical of the business value of hiding any data from the console and selectively revealing it at this point, but on the other hand, if that's what we've decided to do as a company, I think the amount to be withdrawn is more sensitive than pretty much anything else we have in vats at this point.

I'm still catching up. But first, is this simply a misunderstanding? Has this been cleared up? If not, what hiding from the console are we talking about?

Reply:
@dckc the hazards pointed out by the Joe-E paper apply in full force to us. For non-adversarial callees we can mitigate these by following safe patterns. But within the JS-compatibility constraints that SES has taken on, we cannot solve this exception safety issue when Alice and Chuck are malicious conspirators.

The first Caja secured ES3 by whole program rewrite, and a massively intrusive one at that. It was able to solve the exception safety issue. For modern SES, we have to learn to live with the hazard. This will take extra discipline during reviews.

Reply:
The intention of hiding information from the console was my misunderstanding, based on the experience of dealing with information being hidden from the console (inadvertently). I needed to reread the documentation on [assert](https://github.com/endojs/endo/tree/master/packages/ses/src/error). Dan's right that it's about hiding errors from callers, not the console.

I think I'm still trying to understand the exact attack that might happen to us in the short term that we are trying to prevent. I can see an attack given a call stack, which is removed through SES, but I'm not immediately seeing an attack given the actual contents of the substitution values. It seems like the worst-case scenario would be if we put capabilities into the template contents. But do we ever want to do that? Even something like putting an `instance` into the template contents wouldn't be very helpful, because it's an opaque object that has no identity. Is it intended that the errors are read by users with the user's petnames?

Reply:
No, we don't want to put capabilities into error message template contents. We don't want to write evil code like Chuck.

I'm not certain that I understand all of your comment (perhaps we should talk it over), but I approve this PR _without_ use of `q()`.

I was a little surprised that ava's `t.throws()` sees the un-redacted brand info, but I'm glad to see that it does. ava is the sort of debugging tool that _should_ be able to act like the console and see what's going on.

Reply:
Ah... right... I think ava out of the box isn't set up to see un-redacted error details; but we use `prepare-test-env-ava.js`: "SES-Ava can wrap Ava test functions such that errors are reported to the console in full detail. "

Reply:
ses-ava reveals the error in its full detail to the console, but not to the test. The unredacted message in the test case indicates a bug, specifically that it previously used `new Error` instead of `assert`. I think this should not land until the test verifies that the error message does not disclose the amount object. Now that I’m no longer confused, I agree that `X` should be used and `q` should not be used.

Reply:
OK. i un-approved.